src/marker.rs

```diff
@@ -25,153 +364,302 @@ pub fn marker_2_rectancle(
+        let clean_job_guards_ref = guards.clone();
+
+        let cleaner_guard =
+            timer
+                .lock()
+                .unwrap()
+                .schedule_repeating(chrono::Duration::seconds(5), move || {
+                    // every 5s clear the guards that outlived their markers
+                    let mut clean_job_delete_markers = clean_job_delete_markers_ref.lock().unwrap();
+                    let mut clean_job_guards = clean_job_guards_ref.lock().unwrap();
+                    for key in clean_job_delete_markers.iter() {
+                        clean_job_guards.remove(key);
+                    }
+                    clean_job_delete_markers.clear();
+                });
+
+        Self {
+            _markers_container: Arc::new(RwLock::new(marker_container)),
+            _deleted_markers: deleted_markers,
+            _guards: guards,
+            _timer: timer,
+            _cleaner_guard: cleaner_guard,
+        }
+    }
+
+    fn add_marker(&mut self, marker: &rosrust_msg::visualization_msgs::Marker) {
+        self._markers_container.write().unwrap().add_marker(marker);
+
+        // Handle marker lifecycle
+        if marker.lifetime.seconds() == 0.0 {
+            return;
+        }
+
+        let markers_container_ref = self._markers_container.clone();
+
+        let chrono_delay = chrono::Duration::seconds(marker.lifetime.sec as i64)
+            + chrono::Duration::nanoseconds(marker.lifetime.nsec as i64);
+
+        let marker_info = (marker.ns.clone(), marker.id);
+
+        let guard_ = self
+            ._timer
+            .lock()
+            .unwrap()
+            .schedule_with_delay(chrono_delay, move || {
+                let mut mc = markers_container_ref.write().unwrap();
+                mc.delete_marker(marker_info.0.clone(), marker_info.1);
+            });
+
+        let mut guards_ref = self._guards.lock().unwrap();
+        guards_ref.insert((marker.ns.clone(), marker.id), guard_);
+    }
+
+    fn delete_marker(&mut self, marker_ns: String, marker_id: i32) {
+        self._markers_container
+            .write()
+            .unwrap()
+            .delete_marker(marker_ns.clone(), marker_id);
+        self._deleted_markers
+            .lock()
+            .unwrap()
+            .push((marker_ns, marker_id));
+    }
+
+    fn clear(&mut self) {
+        // loose the guards if any, to cancel the timer tasks
+        self._deleted_markers.lock().unwrap().clear();
+        self._guards.lock().unwrap().clear();
+
+        self._markers_container.write().unwrap().clear();
+    }
+
+    fn clear_namespace(&mut self, marker_ns: String) {
+        let removed_ids = self
+            ._markers_container
+            .write()
+            .unwrap()
+            .clear_namespace(marker_ns.clone());
+
+        // schedule deletion of the cleared markers
+        let mut deleted_markers = self._deleted_markers.lock().unwrap();
+        for id in removed_ids {
```

Comment:
Try something like `deleted_markers.extend(removed_ids.iter().map(|id|(marker_ns.clone(),id)).collect());` (didn't test it, you might need some extra stuff)
And you can probably remove the mutability and do it in one shot. But I didn't verify how.

Reply:
Nice idea! Let's call this rustifying code